From 006369fa1dc35f2873eb3151f68dd8e78d42d54a Mon Sep 17 00:00:00 2001
From: Aleksandrs Vinarskis <alex.vinarskis@gmail.com>
Date: Sat, 3 May 2025 13:40:24 +0200
Subject: [PATCH 5/7] fixup: drm/msm/dp: Prepare for link training per-segment
 for LTTPRs

Fetch max VS, PE levels per phy, instead of always assuming level 3
is supported.

Signed-off-by: Aleksandrs Vinarskis <alex.vinarskis@gmail.com>
---
 drivers/gpu/drm/msm/dp/dp_display.c | 8 +++++++-
 drivers/gpu/drm/msm/dp/dp_link.h    | 1 +
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/msm/dp/dp_display.c b/drivers/gpu/drm/msm/dp/dp_display.c
index ab8c1f19dcb4..e61d5348541b 100644
--- a/drivers/gpu/drm/msm/dp/dp_display.c
+++ b/drivers/gpu/drm/msm/dp/dp_display.c
@@ -369,7 +369,7 @@ static int msm_dp_display_send_hpd_notification(struct msm_dp_display_private *d
 
 static int msm_dp_display_lttpr_init(struct msm_dp_display_private *dp, u8 *dpcd)
 {
-	int rc, lttpr_count;
+	int rc, lttpr_count, i;
 
 	if (drm_dp_read_lttpr_common_caps(dp->aux, dpcd, dp->link->lttpr_common_caps))
 		return 0;
@@ -381,6 +381,12 @@ static int msm_dp_display_lttpr_init(struct msm_dp_display_private *dp, u8 *dpcd
 		return 0;
 	}
 
+	for (i = 0; i < lttpr_count; i++) {
+		drm_dp_dump_lttpr_desc(dp->aux, DP_PHY_LTTPR(i));
+		drm_dp_read_lttpr_phy_caps(dp->aux, dpcd, DP_PHY_LTTPR(i),
+					   dp->link->lttpr_phy_caps[i]);
+	}
+
 	return lttpr_count;
 }
 
diff --git a/drivers/gpu/drm/msm/dp/dp_link.h b/drivers/gpu/drm/msm/dp/dp_link.h
index ba47c6d19fbf..9e8b0857a4ce 100644
--- a/drivers/gpu/drm/msm/dp/dp_link.h
+++ b/drivers/gpu/drm/msm/dp/dp_link.h
@@ -62,6 +62,7 @@ struct msm_dp_link_phy_params {
 
 struct msm_dp_link {
 	u8 lttpr_common_caps[DP_LTTPR_COMMON_CAP_SIZE];
+	u8 lttpr_phy_caps[DP_MAX_LTTPR_COUNT][DP_LTTPR_PHY_CAP_SIZE];
 	int lttpr_count;
 
 	u32 sink_request;
-- 
2.45.2

